veps/vep-1.md

```diff
@@ -0,0 +1,100 @@
+# VEP-1 Vitess Release Cycle
+
+```
+VEP: 1
+Title: Vitess Release Cycle
+Author: Morgan Tocker
+Status: Proposed
+Created: 2019-07-22
+```
+
+## Abstract
+
+Vitess releases have historically been infrequent and not on a fixed release schedule. This has led to a status quo of advanced users maintaining their own branches and pulling in changes from Vitess master.
+
+While keeping _master_ in a permanent stable state is desirable, there will be future scenarios where breaking backward compatibility is desired. This VEP aims to codify a release cycle with rules around backward compatibility, supported upgrade paths, and deprecation schedule.
+
+Definitions of *MUST*, *SHOULD*, *MAY* from [RFC2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+## Motivation
+
+The motivation for stable releases is two-fold:
+
+1. Stable releases are easier to consume by novice users. Bug reports, documentation, and support requests will be easier to follow when the user's deployed version is clearer.
+
+2. Advanced users benefit from a clear cycle from which to merge upstream changes into their branch. This also allows for incompatible changes to be introduced at known intervals, which can help improve the health of the project.
+
+## Versioning and Release Cadence
+
+The proposed release cycle is best described as _Monotonic Versioning_ with a release cadence similar to Kubernetes. The terminology `COMPATIBILITY.RELEASE` used in this document is directly sourced from the [Monotonic Versioning Manifesto](http://blog.appliedcompscilab.com/monotonic_versioning_manifesto/):
+
+1. The `COMPATIBILITY` number will be incremented every 12 weeks.
+2. Docker images will be tagged weekly as `COMPATIBILITY.yy+YYYYMMDD`.
+3. Each supported version of Vitess will be tagged weekly. If no changes have been made to the branch from the prior week, a tag will not be created.
+4. Each `COMPATIBILITY` number will be supported by the Open Source Vitess community for 9 months after the initial release. _Support Lifecycle_ is further described below.
+
+Monotonic versioning itself is a modified version of [Semantic Versioning](https://semver.org/).
+
+## Release Criteria
+
+The release cadence is based on time-based rather than feature-based milestones. However; all releases must pass the suite of continuous integration and regression tests prior to release.
+
+Large or compatibility breaking features will be encouraged to merge within the first 6 weeks of the development cycle. Performance testing will be completed on a recurring basis throughout the development cycle.
+
+## Backwards and Forwards Compatibility Promise
+
+For each `COMPATIBILITY` number, both backwards and forwards compatibility **MUST** be ensured. To use hypothetical examples:
```

Comment:
Might be useful to have a way to indicate that a new feature is not yet stable / doesn't yet have the compatibility promise.

Maybe option/flag names like *-experimental_foo* could indicate that a flag is an unstable / in-development feature that allows for the possibility for behavior changes across major releases?

For example, `vtctl` returns wildly inconsistent output formats for all the different commands. I think it would be a good idea to introduce a flag named something like `-experimental-vtctl-json-output` and then slowly update all the different commands to have a json output. That updating process might take weeks or months and span multiple releases. The flag would eventually be renamed to something like `-vtctl-output-format=json` when things are in a state where all commands are able to support json output.

Reply:
(I think of bazel as an example of a tool with a large number of flags that come and go, with a need to clarify which ones are experimental and under what circumstances a flag default can change: https://docs.bazel.build/versions/master/command-line-reference.html)

Reply:
I like this idea very much. I will incorporate it.

Reply:
I have added "exclusions" in this commit: https://github.com/vitessio/enhancements/pull/1/commits/a664ebbf8951e0e4c5daf1d20c98ba138556c60b